src/lenskit/lightfm.py

```diff
@@ -0,0 +1,119 @@
+# This file is part of LensKit.
+# Copyright (C) 2018-2023 Boise State University
+# Copyright (C) 2023-2024 Drexel University
+# Licensed under the MIT license, see LICENSE.md for details.
+# SPDX-License-Identifier: MIT
+
+"""
+LightFM integration with LensKit.
+"""
+
+import logging
+
+import numpy as np
+from lightfm import LightFM
+from pydantic import BaseModel, JsonValue
+from typing_extensions import override
+
+from lenskit.data import Dataset, ItemList, QueryInput, RecQuery, Vocabulary
+from lenskit.pipeline import Component
+from lenskit.training import Trainable, TrainingOptions
+
+_logger = logging.getLogger(__name__)
+
+
+class LightFMConfig(BaseModel, extra="allow"):
+    __pydantic_extra__: dict[str, JsonValue]
+    no_components: int = 50
+    """
+    Number of latent factors to use.
+    """
+    learning_rate: float = 0.05
+    """
+    Learning rate for the LightFM model.
+    """
+    loss: str = "warp"
+    """
+    Loss function to optimize. Options: 'logistic', 'bpr', 'warp', or 'warp-kos'.
+    """
+    epochs: int = 30
+    """
+    Number of training epochs.
+    """
+
+
+class LightFMScorer(Component[ItemList], Trainable):
+    """
+    LightFM scorer for LensKit.
+
+    This uses the LightFM library to implement a hybrid collaborative filtering recommender.
+
+    Stability:
+        Caller
+
+    Args:
+        config:
+            Configuration for the LightFM model.
+        kwargs:
+            Additional arguments to pass to the LightFM constructor.
+    """
+
+    config: LightFMConfig
+
+    users_: Vocabulary
+    items_: Vocabulary
+    user_features_: np.ndarray[tuple[int, int], np.dtype[np.float64]]
+    item_features_: np.ndarray[tuple[int, int], np.dtype[np.float64]]
```

Comment:
Since we are using the model to do predictions, we don't need to save these arrays separately, so let's remove them.